evaluation.md

```diff
@@ -0,0 +1,413 @@
+
+# TAAR – Evaluating existing recommenders
+
+Not every recommender can always make a recommendation. To evaluate the individual recommenders for the ensemble, we want to find out how often this is the case and how well the recommenders complement each other.
+
+This notebook either needs to be executed in the [TAAR](http://github.com/mozilla/taar) repository or somewhere where TAAR is in the Python path, because some TAAR recommenders are loaded in.
+
+## Retrieving the relevant variables from the longitudinal dataset
+
+
+```python
+frame = sqlContext.sql("""
+SELECT client_id,
+    settings[1].locale AS locale,
+    map_keys(active_addons[1]) AS installed_addons,
+    geo_city[1] AS geoCity,
+    subsession_length[1] AS subsessionLength,
+    system_os[1].name AS os,
+    scalar_parent_browser_engagement_total_uri_count[1].value AS total_uri,
+    scalar_parent_browser_engagement_tab_open_event_count[1].value as tab_open_count,
+    places_bookmarks_count[1].sum as bookmark_count,
+    scalar_parent_browser_engagement_unique_domains_count[1].value as unique_tlds
+FROM longitudinal
+""")
+
+rdd = frame.rdd
+```
+
+## Loading addon data (AMO)
+
+We need to load the addon database to find out which addons are legacy addons.
+
+
+```python
+from taar.recommenders.utils import get_s3_json_content
+```
+
+
+```python
+AMO_DUMP_BUCKET = 'telemetry-parquet'
+AMO_DUMP_KEY = 'telemetry-ml/addon_recommender/addons_database.json'
+```
+
+
+```python
+amo_dump = get_s3_json_content(AMO_DUMP_BUCKET, AMO_DUMP_KEY)
+```
+
+## Filtering out legacy addons
+
+This is a helper function that takes a list of addon IDs and only returns the IDs that are from legacy addons.
+
+
+```python
+def get_legacy_addons(installed_addons):
+    legacy_addons = []
+    
+    for addon_id in installed_addons:
+        if addon_id in amo_dump:
+            addon = amo_dump[addon_id]
+            addon_files = addon.get('current_version', {}).get('files', {})
+
+            is_webextension = any([f.get("is_webextension", False) for f in addon_files])
+            is_legacy = not is_webextension
+
+            if is_legacy:
+                legacy_addons.append(addon_id)
+            
+    return legacy_addons
+```
+
+## Completing client data
+
+
+```python
+def complete_client_data(client_data):
+    client = client_data.asDict()
+    
+    client['installed_addons'] = client['installed_addons'] or []
```

Comment:
Here can we filter out legacy addons and system addons? I speculate that the high rate of "can_recommend" positives you see for collaborative recommender is because it expects a more rigorously filtered data source. CollaborativeRecommender().can_recommend is only looking at `len(client_data.get('installed_addons', [])) > 0:`

Can you double check that only valid addons that can be used by collaborative recommender are included in the client sample you pass to it in this notebook.

Reply:
I'm filtering out system addons now. This reduced the number for the collaborative filter to ~40%.